File: codewars/Fun with tree - max sum/maxSum.js
function maxSum(root) {
  
  class TreeNode {
    constructor(value, left = null, right = null) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
}
  
  if (root === null) {
        return 0;
    }

  if (root.left === null && root.right === null) {
        return root.value;
    }
  
  let leftSum, rightSum;
  
  if (root.left !== null){
    leftSum = maxSum(root.left);
  }
  else leftSum = Number.NEGATIVE_INFINITY; // это значение работает корректно для функции Math.max
  
  if (root.right !== null){
    rightSum = maxSum(root.right);
  }
  else rightSum = Number.NEGATIVE_INFINITY;
  
  return root.value + Math.max(leftSum, rightSum); 
}

File: codewars/Fun with tree - max sum/max_sum.py
from preloaded import TreeNode

def max_sum(root: TreeNode) -> int:
    if root is None:
        return 0

    def dfs(node):
        if node is None:
            return float('-inf')

        if node.left is None and node.right is None:
            return node.value

        left_sum = dfs(node.left)
        right_sum = dfs(node.right)

        return node.value + max(left_sum, right_sum)

    return dfs(root)

root = TreeNode(
    5,
    TreeNode(4, TreeNode(-80), TreeNode(-60)),
    TreeNode(10, TreeNode(-90))
)

print(max_sum(root))
